astropy/utils/collections.py

```diff
@@ -56,3 +69,101 @@ def extend(self, x):
         for item in x:
             self._assert(item)
             super().append(item)
+
+
+def _locked_mutator(instance_method):
+    @wraps(instance_method)
+    def inner(self, *args: Any, _owned: bool = False, **kwargs):
+        if not _owned:
+            cls_name = self._owner_class.__name__
+            msg = (
+                f"Directly mutating {cls_name}.{self._attr_name} "
+                f"via {instance_method.__name__} is deprecated and will raise an "
+                "error in the future. "
+            )
+
+            if (repl := self._replacements.get(self._attr_name)) is not None:
+                msg += f"Please use {cls_name}.{repl} instead"
+            else:
+                msg += (
+                    "There is currently no planned replacement for this method. "
+                    "If you need one, please open a feature request at "
+                    "https://github.com/astropy/astropy/issues/new/choose"
+                )
+
+            warnings.warn(
+                msg,
+                category=DeprecationWarning,
+                stacklevel=2,
+            )
+        return instance_method(self, *args, **kwargs)
+
+    return inner
+
+
+class _OwnedList(list):
```

Comment:
It's not intended as a new feature, which is why I'm making it private by default, but I could just as easily make this class public, or move it to a different module, closer to where it's used.